torch_geometric/sampler/neighbor_sampler.py

```diff
@@ -359,103 +340,9 @@ def sample_from_edges(
-                        return_inverse=True)
-                    seed_dst = edge_label_index[1]
-                    seed_dst, inverse_dst = seed_dst.unique(
-                        return_inverse=True)
-                    edge_label_index = torch.stack([
-                        inverse_src,
-                        inverse_dst,
-                    ], dim=0)
-                    seed_dict = {
-                        self.input_type[0]: seed_src,
-                        self.input_type[-1]: seed_dst,
-                    }
-
-            else:  # Merge both source and destination node indices:
-                if self.disjoint:
-                    seed_nodes = edge_label_index.view(-1)
-                    edge_label_index = torch.arange(0, 2 * num_seed_edges)
-                    edge_label_index = edge_label_index.view(2, -1)
-                    seed_dict = {self.input_type[0]: seed_nodes}
-                    if edge_label_time is not None:
-                        tmp = torch.cat([edge_label_time, edge_label_time])
-                        seed_time_dict = {self.input_type[0]: tmp}
-                else:
-                    seed_nodes = edge_label_index.view(-1)
-                    seed_nodes, inverse = seed_nodes.unique(
-                        return_inverse=True)
-                    edge_label_index = inverse.view(2, -1)
-                    seed_dict = {self.input_type[0]: seed_nodes}
-
-            output = self._sample(
-                seed=seed_dict,
-                seed_time_dict=seed_time_dict,
-            )
-
-            if self.disjoint:
-                for key, batch in output.batch.items():
-                    output.batch[key] = batch % num_seed_edges
-
-            output.metadata = (index, edge_label_index, edge_label,
-                               edge_label_time)
-
-        elif issubclass(self.data_cls, Data):
-            if self.disjoint:
-                seed_nodes = edge_label_index.view(-1)
-                edge_label_index = torch.arange(0, 2 * num_seed_edges)
-                edge_label_index = edge_label_index.view(2, -1)
-                if edge_label_time is not None:
-                    seed_time = torch.cat([edge_label_time, edge_label_time])
-
-            else:
-                seed_nodes = edge_label_index.view(-1)
-                seed_nodes, inverse = seed_nodes.unique(return_inverse=True)
-                edge_label_index = inverse.view(2, -1)
-
-            output = self._sample(seed=seed_nodes, seed_time=seed_time)
-
-            if self.disjoint:
-                output.batch = output.batch % num_seed_edges
-
-            output.metadata = (index, edge_label_index, edge_label,
-                               edge_label_time)
-
-        else:
-            raise TypeError(f"'{self.__class__.__name__}'' found invalid "
-                            f"type: '{self.data_cls}'")
-
-        return output
+        return edge_sample(index, self._sample, self.num_src_nodes,
```

Comment:
Do we need to move the implementation to `edge_sample` and `node_sample`. Can't we keep them here and have classes inherit from `NeighborSampler`?

Reply:
I don't think it's easy for classes to inherit from NeighborSampler, since NeighborSampler makes a lot of assumptions about in-memory sampling that folks may want to avoid when writing a custom neighbor sampler. If we want to make NeighborSampler a more well-defined base class, I am open to it, but for now I think it makes more sense to just separate things out.